File: examples/online-boutique/mobile/lib/api/payment.dart
// ignore_for_file: unused_import
library payment;

import 'dart:convert';
import 'dart:typed_data';
import 'ftl_client.dart';
import 'builtin.dart' as builtin;
import 'currency.dart' as currency;


class ChargeRequest{
  currency.Money amount;
  CreditCardInfo creditCard;

  ChargeRequest({  required this.amount,  required this.creditCard,  });

  Map<String, dynamic> toJson() {
    return {
      'amount': ((dynamic v) => v.toJson())(amount),
      'creditCard': ((dynamic v) => v.toJson())(creditCard),
    };
  }

  factory ChargeRequest.fromJson(Map<String, dynamic> map) {
    return ChargeRequest(
      amount: ((dynamic v) => currency.Money.fromJson(v))(map['amount']), creditCard: ((dynamic v) => CreditCardInfo.fromJson(v))(map['creditCard']), 
    );
  }
}

class ChargeResponse{
  String transactionId;

  ChargeResponse({  required this.transactionId,  });

  Map<String, dynamic> toJson() {
    return {
      'transactionId': ((dynamic v) => v)(transactionId),
    };
  }

  factory ChargeResponse.fromJson(Map<String, dynamic> map) {
    return ChargeResponse(
      transactionId: ((dynamic v) => v)(map['transactionId']), 
    );
  }
}

class CreditCardInfo{
  String number;
  int cvv;
  int expirationYear;
  int expirationMonth;

  CreditCardInfo({  required this.number,  required this.cvv,  required this.expirationYear,  required this.expirationMonth,  });

  Map<String, dynamic> toJson() {
    return {
      'number': ((dynamic v) => v)(number),
      'cvv': ((dynamic v) => v)(cvv),
      'expirationYear': ((dynamic v) => v)(expirationYear),
      'expirationMonth': ((dynamic v) => v)(expirationMonth),
    };
  }

  factory CreditCardInfo.fromJson(Map<String, dynamic> map) {
    return CreditCardInfo(
      number: ((dynamic v) => v)(map['number']), cvv: ((dynamic v) => v)(map['cvv']), expirationYear: ((dynamic v) => v)(map['expirationYear']), expirationMonth: ((dynamic v) => v)(map['expirationMonth']), 
    );
  }
}

class ErrorResponse{
  String message;

  ErrorResponse({  required this.message,  });

  Map<String, dynamic> toJson() {
    return {
      'message': ((dynamic v) => v)(message),
    };
  }

  factory ErrorResponse.fromJson(Map<String, dynamic> map) {
    return ErrorResponse(
      message: ((dynamic v) => v)(map['message']), 
    );
  }
}


class PaymentClient {
  final FTLHttpClient ftlClient;

  PaymentClient({required this.ftlClient});


  Future<ChargeResponse> charge(
    ChargeRequest request, { 
    Map<String, String>? headers,
  }) async {
    final response = await ftlClient.post('/payment/charge', request: request.toJson());
    if (response.statusCode == 200) {
      final body = json.decode(utf8.decode(response.bodyBytes));
      return ChargeResponse.fromJson(body);
    } else {
      throw Exception('Failed to get charge response');
    }
  }
}

File: examples/online-boutique/backend/services/payment/payment.go
//ftl:module payment
package payment

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"

	"ftl/currency"

	"ftl/builtin"

	"github.com/TBD54566975/ftl/go-runtime/ftl"
)

type ErrorResponse struct {
	Message string `json:"message"`
}

type CreditCardInfo struct {
	Number          string
	CVV             int
	ExpirationYear  int
	ExpirationMonth int
}

// LastFour returns the last four digits of the card number.
func (c CreditCardInfo) LastFour() string {
	num := c.Number
	if len(num) > 4 {
		num = num[len(num)-4:]
	}
	return num
}

type ChargeRequest struct {
	Amount     currency.Money
	CreditCard CreditCardInfo
}

type ChargeResponse struct {
	TransactionID string
}

//ftl:ingress POST /payment/charge
func Charge(ctx context.Context, req builtin.HttpRequest[ChargeRequest, ftl.Unit, ftl.Unit]) (builtin.HttpResponse[ChargeResponse, ErrorResponse], error) {
	card := req.Body.CreditCard
	number := strings.ReplaceAll(card.Number, "-", "")
	var company string
	switch {
	case len(number) < 4:
		return builtin.HttpResponse[ChargeResponse, ErrorResponse]{Error: ftl.Some(ErrorResponse{Message: "Invalid card number"})}, nil
	case number[0] == '4':
		company = "Visa"
	case number[0] == '5':
		company = "MasterCard"
	default:
		return builtin.HttpResponse[ChargeResponse, ErrorResponse]{Error: ftl.Some(ErrorResponse{Message: "Invalid card number"})}, nil
	}
	if card.CVV < 100 || card.CVV > 9999 {
		return builtin.HttpResponse[ChargeResponse, ErrorResponse]{Error: ftl.Some(ErrorResponse{Message: "Invalid CVV number"})}, nil
	}
	if time.Date(card.ExpirationYear, time.Month(card.ExpirationMonth), 0, 0, 0, 0, 0, time.Local).Before(time.Now()) {
		return builtin.HttpResponse[ChargeResponse, ErrorResponse]{Error: ftl.Some(ErrorResponse{Message: "Card expired"})}, nil
	}

	// Card is valid: process the transaction.
	fmt.Println(
		"Transaction processed",
		"company", company,
		"last_four", number[len(number)-4:],
		"currency", req.Body.Amount.CurrencyCode,
		"amount", fmt.Sprintf("%d.%d", req.Body.Amount.Units, req.Body.Amount.Nanos),
	)
	return builtin.HttpResponse[ChargeResponse, ErrorResponse]{
		Body: ftl.Some(ChargeResponse{TransactionID: uuid.New().String()}),
	}, nil
}
